### PR TITLE
Add support for the DeviceLocalLayout API when lowering FFI calls.

### PR DESCRIPTION
**Pull request by dfm**
This PR updates the FFI lowering rule to support a DeviceLoweringLayout object as input when specifying the input and output layouts. For now, this just converts the DLL object to its appropriate list of minor-to-major integers because that's what the custom call op expects.

### CODE REVIEW

**Review by yashk2810 — APPROVED** [review]
Thanks!

File: jax/_src/extend/ffi.py [hunk 1]
```diff
@@ -104,15 +106,27 @@ def _aval_shape(aval: core.AbstractValue) -> Shape:
   return () if aval is core.abstract_token else aval.shape  # pytype: disable=attribute-error
 
 
-def _default_layouts(avals: Iterable[core.AbstractValue]) -> list[list[DimSize]]:
-  return [list(reversed(range(len(_aval_shape(aval))))) for aval in avals]
+def _default_layout(aval: core.AbstractValue) -> list[int]:
```
[the remark this answers is not in the document]
Comment (dfm) [reply jax/_src/extend/ffi.py]: Good idea - done!

File: jax/_src/extend/ffi.py [hunk 2]
```diff
@@ -27,12 +27,14 @@
...
 from jax._src.interpreters import mlir
+from jax._src.layout import DeviceLocalLayout
 from jax._src.lib import jaxlib
 from jax._src.lib import xla_client
 from jax._src.lib.mlir import ir
-from jax._src.typing import Array, ArrayLike, DimSize, DuckTypedArray, Shape
+from jax._src.typing import Array, ArrayLike, DuckTypedArray, Shape
 
 map, unsafe_map = util.safe_map, map
+FfiLayoutOptions = Sequence[int] | DeviceLocalLayout | None
```
[the remark this answers is not in the document]
Comment (yashk2810) [reply jax/_src/extend/ffi.py:37]: Good point but a lot of our APIs are in experimental lol (shard_map being a prime one). We can take it out of experimental but would be nice to lay the foundation to move to that API!

[the remark this answers is not in the document]
Comment (dfm) [reply jax/_src/extend/ffi.py:37]: haha yeah, I think we might have commitment issues! 🤣

**Pull request merged**